src/devices/threads.rs

```diff
@@ -238,6 +243,70 @@ pub struct DeviceSet {
     event_manager: SubscriberEventManager,
 }
 
+fn ioregion_event_loop(
+    should_stop: &Arc<AtomicBool>,
+    device_ready: &Arc<DeviceReady>,
+    mmio_mgr: Arc<Mutex<IoPirate>>,
+    device: Arc<Mutex<dyn MaybeIoRegionFd + Send>>,
+) -> Result<()> {
+    let ioregionfd = {
+        let device = try_with!(device.lock(), "cannot lock device");
+        let ioregionfd = device.get_ioregionfd();
+        ioregionfd.ok_or(simple_error!(
+            "cannot start ioregion event loop when ioregion does not exist"
+        ))?
+    };
+    device_ready.notify()?;
+
+    loop {
+        let cmd = try_with!(
+            ioregionfd.read(),
+            "cannot read mmio command from ioregionfd (fd {:?})",
+            ioregionfd
+        );
+        {
+            let mut mmio_mgr = try_with!(
+                mmio_mgr.lock(),
+                "cannot lock mmio manager to handle mmio command"
+            );
+            mmio_mgr.handle_ioregion_rw(&ioregionfd, cmd)?;
+        }
+
+        if should_stop.load(Ordering::Relaxed) {
+            break;
+        }
+    }
+    Ok(())
+}
+
+/// see handle_mmio_exits
+fn ioregion_handler_thread(
+    device: Arc<Mutex<dyn MaybeIoRegionFd + Send>>,
+    mmio_mgr: Arc<Mutex<IoPirate>>,
+    err_sender: &SyncSender<()>,
+    device_ready: &Arc<DeviceReady>,
+) -> Result<InterrutableThread<()>> {
+    let device_ready = device_ready.clone();
+
+    let res = InterrutableThread::spawn(
+        "ioregion-handler",
+        err_sender,
+        move |should_stop: Arc<AtomicBool>| {
+            info!("ioregion mmio handler started");
+            try_with!(
+                ioregion_event_loop(&should_stop, &device_ready, mmio_mgr, device),
+                "foo"
+            );
+
+            // TODO drop remote resources like ioeventfd before disowning traced process?
```

Comment:
Does this currently leak file descriptors?

Reply:
yes. See PR description. Also it doesn't because ioregionfd code is disabled for now.